### PR TITLE
Answers init returns a promise

### PR DESCRIPTION
**Pull request by cea2aj**
This PR changes ANSWERS.init() such that it returns a promise which resolves when initialization is complete. Both the onReady callback and the promise may be used. If both are included, the onReady callback is called first, and then the promise is resolved.

Usage looks like this:
```javascript
ANSWERS.init({
  //config
}).then(() => {
  //onReady code 
});
```

This commit also changes _handlePonyfillCssVariables such that it returns a promise.

J=SLAP-504
TEST=manual

I tested this with html that initializes the SearchBar and Universal results. I tested onReady code just in the callback, in both the callback and in the promise, and in just the promise. I confirmed onReady executes before the promise resolves.

I tested the code with MasterSwitchAPI and with ponyfillCss manually disabled.

I also tested on IE11 and confirmed that the existing polylfills work so long as arrow functions are not used.

### CODE REVIEW

**Comment by cea2aj** [comment]
We could clean up this code a bit more by removing _handlePonyfillCssVariables altogether if we simply call await ponyfillCssVariables inside the init, but we would need to make the init async, and I'm not sure if that's a good idea

File: src/answers-umd.js [hunk 1]
```diff
@@ -259,52 +259,28 @@ class Answers {
...
     }
 
-    // Templates are currently downloaded separately from the CORE and UI bundle.
-    // Future enhancement is to ship the components with templates in a separate bundle.
-    this.templates = new DefaultTemplatesLoader(templates => {
-      this.renderer.init(templates);
-      this._handlePonyfillCssVariables(
-        parsedConfig.disableCssVariablesPonyfill,
-        this._invokeOnReady.bind(this));
+    this.promise = new Promise((resolve, reject) => {
```
Comment (creotutar) [inline src/answers-umd.js]: Does this stop us from being backwards compatible? It seems like people _must_ use the Promise with these changes. Is this what we wanted, considering a lot of current customers are using a non-promised version?

Comment (cea2aj) [reply src/answers-umd.js]: Nope, it's still backwards compatible with onReady in the config. Right before the promise is resolved onReady is executed, even you if aren't using the promise version.

Comment (creotutar) [reply src/answers-umd.js]: Love the new promises around css-vars-ponyfill

Comment (oshi97) [inline src/answers-umd.js]: This is definitely a creative solution! I would still rather just return the promise directly, but if Rose thinks we need to keep returning the ANSWERS instance then this looks good to me. I thought about maybe naming it onReadyPromise but that doesn't seem as elegant for some reason?

File: src/answers-umd.js [hunk 2]
```diff
@@ -259,50 +259,29 @@ class Answers {
...
-    // Future enhancement is to ship the components with templates in a separate bundle.
-    this.templates = new DefaultTemplatesLoader(templates => {
-      this.renderer.init(templates);
-      this._handlePonyfillCssVariables(
-        parsedConfig.disableCssVariablesPonyfill,
-        this._invokeOnReady.bind(this));
+    const onReadyPromise = new Promise((resolve, reject) => {
+      this._handlePonyfillCssVariables(parsedConfig.disableCssVariablesPonyfill)
+        .then(this._masterSwitchApi.isDisabled())
+        .then((isDisabled) => {
```
Comment (oshi97) [inline src/answers-umd.js]: nit but I think our style is to normally omit the parenthesis around a javascript arrow function'
s param if there's only one param

File: src/answers-umd.js [hunk 3]
```diff
@@ -259,50 +259,29 @@ class Answers {
...
+          if (!isDisabled) {
```
Comment (oshi97) [inline src/answers-umd.js]: do we need to reject if isDisabled is true?

File: src/answers-umd.js (hunk 2 above)
Comment (cea2aj) [reply src/answers-umd.js]: Will do!

File: src/answers-umd.js [hunk 4]
```diff
@@ -259,50 +259,29 @@ class Answers {
...
     }
 
-    // Templates are currently downloaded separately from the CORE and UI bundle.
-    // Future enhancement is to ship the components with templates in a separate bundle.
-    this.templates = new DefaultTemplatesLoader(templates => {
-      this.renderer.init(templates);
-      this._handlePonyfillCssVariables(
-        parsedConfig.disableCssVariablesPonyfill,
-        this._invokeOnReady.bind(this));
+    const onReadyPromise = new Promise((resolve, reject) => {
```
Comment (oshi97) [inline src/answers-umd.js]: oh hmm sorry I missed this during the first round but could we just return the Promise returned by _handlePonyfillCssVariables? instead of wrapping it inside a new promise

File: src/answers-umd.js (hunk 3 above)
Comment (cea2aj) [reply src/answers-umd.js]: I didn't call reject because I was trying to match the original logic flow as closely as possible, however I think you're right. It should be called.

File: src/answers-umd.js (hunk 4 above)
Comment (cea2aj) [reply src/answers-umd.js]: I wrapped it in a promise because the init resolving depends on the results of _handlePonyfillCssVariables promise and the MasterSwitchApi.isDisabled(), which is also a promise. If you can simplify it, I'm happy to take a look

Comment (cea2aj) [reply src/answers-umd.js]: Update: We managed to simplify this!

File: src/answers-umd.js [hunk 5]
```diff
@@ -452,33 +427,57 @@ class Answers {
...
   }
 
+  /**
+   * A promise that resolves when ponyfillCssVariables resolves,
+   * or resolves immediately if ponyfill is disabled
+   * @param {boolean} option to opt out of the css variables ponyfill
+   * @return {Promise} resolves after ponyfillCssVariables, or immediately if disabled
+   */
+  _handlePonyfillCssVariables (ponyfillDisabled) {
+    return new Promise((resolve, reject) => {
```
Comment (oshi97) [inline src/answers-umd.js:440]: nit but I think we can shorten this one slightly too, by doing
```js
if (!ponyfillDisabled) {
  return Promise.resolve();
}
return this.ponyfillCssVariables()
```
Another option, though, might be to have ponyfillCssVariables() go back to just being a pass through to cssVars(), and have all of the promise logic go into this _handle function? I think that divides up responsibility a little more neatly (let me know if you have any other thoughts). Then this _handle method could return a Promise.resolve() if ponyfills are disabled, or a new Promise to passes resolve into the onFinally of ponyfillCssVariables if not disabled

Comment (cea2aj) [reply src/answers-umd.js:440]: I like that idea. It's definitely a little cleaner

File: README.md [hunk 6]
```diff
@@ -98,7 +116,8 @@ function initAnswers() {
     apiKey: '<API_KEY_HERE>',
     // Required, the key used for your Answers experience
     experienceKey: '<EXPERIENCE_KEY_HERE>',
-    // Required, initialize components here, invoked when the Answers component library is loaded/ready
+    // Required, initialize components here, invoked when the Answers component library is loaded/ready.
+    //   Components may instead be initialized after the init promise resolves.
```
Comment (oshi97) [inline README.md]: since this isn't required anymore, remove the "Required" and maybe reword this slightly

File: src/answers-umd.js [hunk 7]
```diff
@@ -452,6 +427,23 @@ class Answers {
...
+   * @param {boolean} option to opt out of the css variables ponyfill
+   * @return {Promise} resolves after ponyfillCssVariables, or immediately if disabled
+   */
+  _handlePonyfillCssVariables (ponyfillDisabled) {
+    if (ponyfillDisabled) {
+      return Promise.resolve();
+    }
+    return new Promise((resolve, reject) => {
+      this.ponyfillCssVariables({
+        onFinally: resolve()
```
Comment (oshi97) [inline src/answers-umd.js]: here, we want to pass in a function. currently what this does (I believe at least) is execute the resolve as this line as evaluated, and pass the return value of resolve, as oppose to pass the resolve function itself into ponyfilCssVariables

Comment (oshi97) [reply src/answers-umd.js]: either just resolve or () => resolve() should work (assuming I'm not bonkers)

Comment (cea2aj) [reply src/answers-umd.js]: Ahh, good catch. Just changed it to resolve.

File: README.md [hunk 8]
```diff
@@ -98,7 +116,8 @@ function initAnswers() {
     apiKey: '<API_KEY_HERE>',
     // Required, the key used for your Answers experience
     experienceKey: '<EXPERIENCE_KEY_HERE>',
-    // Required, initialize components here, invoked when the Answers component library is loaded/ready
+    // Optional, initialize components here, invoked when the Answers component library is loaded/ready.
+    //    If components are not added here, they must be added when the init promise resolves
```
Comment (oshi97) [inline README.md]: definitely just my opinion but "must be added" feels a little too strong for me here. technically, you can also add components in both places and things will work fine. Maybe just "they can also" instead of "they must be"?

Comment (cea2aj) [reply README.md]: Sure, that sounds good

**Review by oshi97 — APPROVED** [review]
lgtm!
